File: Java/947.most-stones-removed-with-same-row-or-column.java
/*
 * @lc app=leetcode id=947 lang=java
 *
 * [947] Most Stones Removed with Same Row or Column
 */

// @lc code=start
class Solution {
    public int removeStones(int[][] stones) {
        if (stones == null || stones.length == 0) return 0;
        Set<String> visited = new HashSet<>();
        int count = 0;
        for (int[] stone : stones) {
            if (visited.add(stone[0] + "-" + stone[1])) {
                dfs(stones, visited, stone);
                count++;
            }
        }
        return stones.length - count;
    }

    private void dfs(int[][] stones, Set<String> visited, int[] stone) {
        for (int[] next : stones) {
            if ((next[0] == stone[0] || next[1] == stone[1]) && (visited.add(next[0] + "-" + next[1]))) {
                dfs(stones, visited, next);
            }
        }
    }
}
// @lc code=end

File: Go/947.most-stones-removed-with-same-row-or-column.go
/*
 * @lc app=leetcode id=947 lang=golang
 *
 * [947] Most Stones Removed with Same Row or Column
 */

// @lc code=start
func removeStones(stones [][]int) int {
    u := newUnion(20000)
    for _, s := range stones {
        u.union(s[0], s[1] + 10000)     // 避免坐标轴的重叠，y坐标加10000
    }

    keeps := make(map[int]int, 1000)
    for _, s := range stones {
        root := u.find(s[0])
        keeps[root]++
    }
    return len(stones) - len(keeps)
}

type union struct {
    parent []int
}

func newUnion(size int) *union {
    parent := make([]int, size)
    for i := range parent {
        parent[i] = i
    }
    return &union {
        parent : parent,
    }
}

func (u *union) union(x, y int) {
    xp, yp := u.find(x), u.find(y)
    u.parent[yp] = xp
}

func (u *union) find(x int) int {
    for u.parent[x] != x {
        u.parent[x] = u.parent[u.parent[x]]
        x = u.parent[x]
    }
    return x
}
// @lc code=end

